pkg/reconciler/sequence/sequence.go

```diff
@@ -331,3 +338,155 @@ func (r *Reconciler) removeUnwantedSubscriptions(ctx context.Context, seq *v1.Se
+	}
+
+	// Prepare maps for lookups
+	existingPolicyMap := make(map[string]*eventingv1alpha1.EventPolicy)
+	for _, policy := range existingPolicies {
+		existingPolicyMap[policy.Name] = policy
+	}
+
+	// Prepare lists for different actions
+	var policiesToUpdate, policiesToCreate, policiesToDelete []*eventingv1alpha1.EventPolicy
+
+	// Handle intermediate channel policies (skip the first channel as it's the input channel!)
+	for i := 1; i < len(channels); i++ {
+		expectedPolicy := resources.MakeEventPolicyForSequenceChannel(s, channels[i], subs[i-1])
+		existingPolicy, exists := existingPolicyMap[expectedPolicy.Name]
+
+		if exists {
+			if !equality.Semantic.DeepEqual(existingPolicy.Spec, expectedPolicy.Spec) {
+				policiesToUpdate = append(policiesToUpdate, expectedPolicy)
+			}
+			delete(existingPolicyMap, expectedPolicy.Name)
+		} else {
+			policiesToCreate = append(policiesToCreate, expectedPolicy)
+		}
+	}
+
+	// Handle input channel policy
+	inputPolicy, err := r.prepareInputChannelEventPolicy(ctx, s, channels[0])
+	if err != nil {
+		return fmt.Errorf("failed to prepare input channel EventPolicy: %w", err)
+	}
+	if inputPolicy != nil {
+		// The sequence has the event policy, so we are creating the event policy for the input channel
+		existingInputPolicy, exists := existingPolicyMap[inputPolicy.Name]
+		if exists {
+			if !equality.Semantic.DeepEqual(existingInputPolicy.Spec, inputPolicy.Spec) {
+				policiesToUpdate = append(policiesToUpdate, inputPolicy)
+			}
+			delete(existingPolicyMap, inputPolicy.Name)
+		} else {
+			policiesToCreate = append(policiesToCreate, inputPolicy)
+		}
+	}
+
+	// Any remaining policies in the map should be deleted
+	for _, policy := range existingPolicyMap {
+		policiesToDelete = append(policiesToDelete, policy)
+	}
+
+	// Perform the actual CRUD operations
+	if err := r.createEventPolicies(ctx, policiesToCreate); err != nil {
+		return fmt.Errorf("failed to create EventPolicies: %w", err)
+	}
+	if err := r.updateEventPolicies(ctx, policiesToUpdate); err != nil {
+		return fmt.Errorf("failed to update EventPolicies: %w", err)
+	}
+	if err := r.deleteEventPolicies(ctx, policiesToDelete); err != nil {
+		return fmt.Errorf("failed to delete EventPolicies: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) listEventPoliciesForSequence(ctx context.Context, s *v1.Sequence) ([]*eventingv1alpha1.EventPolicy, error) {
+	labelSelector := labels.SelectorFromSet(map[string]string{
+		resources.SequenceChannelEventPolicyLabelPrefix + "sequence-name": s.Name,
+	})
+	return r.eventPolicyLister.EventPolicies(s.Namespace).List(labelSelector)
+}
+
+func (r *Reconciler) prepareInputChannelEventPolicy(ctx context.Context, s *v1.Sequence, inputChannel *eventingduckv1.Channelable) (*eventingv1alpha1.EventPolicy, error) {
+	// Trying to see whether the user manually created the eventpolicy for the sequence
+	sequencePolicy, err := r.eventPolicyLister.EventPolicies(s.Namespace).Get(s.Name + "-ep")
```

Comment:
Here we are trying to fetch the eventpolicy by name, what if user creates an event policy with different name ?

Reply:
Great point, multiple event policies can be bound to a Sequence (using the To), we need to find all the matching ones and propagate them down to the input channel

Reply:
I think we can use the `auth.GetEventPoliciesForResource` function